src/util/indexing.jl

```diff
@@ -0,0 +1,150 @@
+module Indexing
+
+export
+    Indexer, names!, rename!, rename
+
+
+intsAndVecs = Union{Unsigned,
+                    AbstractVector{UInt8},
+                    AbstractVector{UInt16},
+                    AbstractVector{UInt32},
+                    AbstractVector{UInt64},
+                    AbstractVector{UInt128}}
+
+# The indexer Type
+type Indexer{T <: intsAndVecs}
+    lookup::Dict{Symbol, T}
+    names::Vector{Symbol}
+end
+
+# Conveinience constructors for creating Indexer that associate one name,
+# with in integer value.
+function Indexer{A <: AbstractVector, T <: Unsigned}(names::A{Symbol}, ::Type{T})
+    u = make_unique(names)
+    lookup = Dict{Symbol, Unsigned}(zip(u, UnitRange{T}(1, length(u))))
+    Indexer{T}(lookup, u)
+end
+function Indexer{S <: AbstractString, T <: Unsigned}(names::Vector{S}, ::Type{T})
+    Indexer(convert(Vector{Symbol}, names), T)
+end
+
+# Conveinience constructors for creating Indexer that associate one name,
+# with several integer values.
+function Indexer{A <: AbstractVector, E <: Unsigned}(names::Vector{Symbol}, groups::A{E})
+    @assert length(names) == length(groups)
+    u = make_unique(names)
+    lookup = Dict{Symbol, A{E}}(zip(u, groups))
+    Indexer(lookup, u)
+end
+function Indexer{S <: AbstractString, A <: AbstractVector, E <: Unsigned}(names::Vector{S}, groups::A{E})
+    Indexer(convert(Vector{Symbol}, names), groups)
+end
+function Indexer{T <: intsAndVecs}(::Type{T})
+    Indexer(Dict{Symbol, T}(), Symbol[])
+end
+
+function make_unique(names::Vector{Symbol})
+    seen = Set{Symbol}()
+    names = copy(names)
+    dups = Int[]
+    for i in 1:length(names)
```

Comment:
Good catch, although I think one second thoughts, the i can be done away with in both loops in the make_unique function and just name be used instead.